[view-transitions] Suppress rendering while a view-transition is being setup.

mattwoodrow opened the pull request:
#### 365b5b3b07b0ccc3be56f0a7bcce3f912f5d5b27
<pre>
[view-transitions] Suppress rendering while a view-transition is being setup.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270672">https://bugs.webkit.org/show_bug.cgi?id=270672</a>
<a href="https://rdar.apple.com/124240939">rdar://124240939</a>

Reviewed by Simon Fraser.

This implements rendering suppression by adding a flag to GraphicsLayer that
prevents updates to the PlatformCALayer tree for it and its descendants until
the flag is cleared. This will currently only work on cocoa, with remote
layers enabled.

The flag gets set on the GraphicsLayer after the rendering update completes,
so that any mutations made before rendering suppression were enabled (including
layer tree changes needed for suppression) get flushed before blocking all
future changes.

It also moves all consumers of RenderStyle:::usedPointerEvents to go via the
renderer, where we can override them to none while suppression is in place.
The exceptions are the style change handlers, which can remain unchanged
since we invalidate style entirely when the VT starts.

This doesn&apos;t yet prevent async scrolling rendering updates - bug 273612.
It also doesn&apos;t work fully with in-process layers - bug 273613.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/input-targets-root-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-css-animation-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-raf-while-render-blocked-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setActiveViewTransition):
(WebCore::Document::renderingIsSuppressedForViewTransition const):
(WebCore::Document::setRenderingIsSuppressedForViewTransitionAfterUpdateRendering):
(WebCore::Document::clearRenderingIsSuppressedForViewTransition):
(WebCore::Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges):
(WebCore::Document::performPendingViewTransitions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::setupViewTransition):
(WebCore::ViewTransition::activateViewTransition):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::extractTargets):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::Page::doAfterUpdateRendering):
(WebCore::Page::didCompleteRenderingFrame):
(WebCore::Page::prioritizeVisibleResources):
(WebCore::Page::forEachRenderableDocument const):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::renderingIsSuppressedIncludingDescendants const):
(WebCore::GraphicsLayer::setRenderingIsSuppressedIncludingDescendants):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::needsCommit):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerRenderingIsSuppressedIncludingDescendants const):
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::unite):
(WebCore::EventRegion::unite):
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::visibleToHitTesting const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::setRenderingIsSuppressed):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::usedPointerEvents const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::checkIntersection):
(WebCore::RenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::nodeAtPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::checkIntersection):
(WebCore::LegacyRenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::nodeAtFloatPoint):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::recursiveMarkWillBeDisplayed):
(WebKit::PlatformCALayerRemote::recursiveBuildTransaction):

Canonical link: <a href="https://commits.webkit.org/279857@main">https://commits.webkit.org/279857@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0abd9a2314a3b52459f33e2a0363d841b679fc0b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34192 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58043 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57065 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41739 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5509 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44356 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3712 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32324 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47442 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25480 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4780 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3637 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4997 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59633 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30010 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51778 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47525 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51181 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12024 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32159 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30939 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-white-flash-before-activation.html:
```diff
@@ -20,7 +20,7 @@
         window.addEventListener("load", () => {
             document.startViewTransition(() => {
-                requestAnimationFrame(() => document.documentElement.classList.remove("reftest-wait"));
+                setTimeout(() => document.documentElement.classList.remove("reftest-wait"), 0);
```
Should no longer be needed after https://github.com/WebKit/WebKit/pull/26817

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -986,6 +986,18 @@ bool LocalFrameView::flushCompositingStateForThisFrame(const LocalFrame& rootFra
 
+void LocalFrameView::enableRenderingSuppressionStateForThisFrame()
+{
+    RenderView* renderView = this->renderView();
```
```suggestion
    auto* renderView = this->renderView();
```

or CheckedPtr

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1186,6 +1198,22 @@ bool LocalFrameView::flushCompositingStateIncludingSubframes()
+{
+    enableRenderingSuppressionStateForThisFrame();
+
+    for (Frame* child = m_frame->tree().firstRenderedChild(); child; child = child->tree().traverseNextRendered(m_frame.ptr())) {
```
```suggestion
    for (auto* child = m_frame->tree().firstRenderedChild(); child; child = child->tree().traverseNextRendered(m_frame.ptr())) {
```

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3955,6 +3965,25 @@ void Page::forEachDocument(const Function<void(Document&)>& functor) const
+{
+    Vector<Ref<Document>> documents;
+    const Frame& mainFrame = protectedMainFrame();
+    for (const Frame* frame = &mainFrame; frame; frame = frame->tree().traverseNext()) {
```
```suggestion
    const auto& mainFrame = protectedMainFrame();
    for (const auto* frame = &mainFrame; frame; frame = frame->tree().traverseNext()) {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked-expected.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<title>View transitions: Rendering suppression prevents painting (ref)</title>
+<link rel="help" href="https://github.com/WICG/view-transitions">
```
The help link can be dropped from the ref. (Side note: you might want to have the `-ref.html` file as well in this patch to export easily)

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked.html:
```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE html>
+<html class=reftest-wait>
+<title>View transitions: Rendering suppression prevents painting</title>
+<link rel="help" href="https://github.com/WICG/view-transitions">
```
https://drafts.csswg.org/css-view-transitions-1/#document-rendering-suppression-for-view-transitions would probably be a better link.

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -798,6 +801,7 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     bool m_userInteractionEnabled : 1;
     bool m_canDetachBackingStore : 1;
     bool m_shouldPaintUsingCompositeCopy : 1;
+    bool m_renderingSuppressed : 1;
```
Bitfield entries can use initializers these days.

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
```diff
@@ -697,6 +701,7 @@ class GraphicsLayerCA : public GraphicsLayer, public PlatformCALayerClient {
     bool m_hasEverPainted : 1;
     bool m_hasDescendantsWithRunningTransformAnimations : 1;
     bool m_hasDescendantsWithUncommittedChanges : 1;
+    bool m_hasAncestorWithRenderingSuppression : 1;
```
Bitfield entries can use initializers these days.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1186,6 +1198,22 @@ bool LocalFrameView::flushCompositingStateIncludingSubframes()
     return allFramesFlushed;
 }
 
+void LocalFrameView::enableRenderingSuppressionStateIncludingSubframes()
```
Move this function up to be below `enableRenderingSuppressionStateForThisFrame()`

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1186,6 +1198,22 @@ bool LocalFrameView::flushCompositingStateIncludingSubframes()
+        auto* localChild = dynamicDowncast<LocalFrame>(child);
+        if (!localChild)
+            continue;
+        auto* frameView = localChild->view();
```
strong pointer

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1861,14 +1861,19 @@ void Page::updateRendering()
             frame->checkedLoader()->restoreScrollPositionAndViewStateNowIfNeeded();
     });
 
+    // FIXME: This updates the scrolling tree (for in-process layers, maybe not
```
I'm not sure how important this FIXME is. Maybe it justifies filing a bug?

smfr commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2384,6 +2384,15 @@ void RenderObject::RepaintRects::transform(const TransformationMatrix& matrix, f
 
+PointerEvents RenderObject::usedPointerEvents() const
+{
+    // FIXME: This could be a performance problem, checking this all
```
`renderingSuppressedForViewTransition()` is just checking a bit so it's probably fine?

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10499,6 +10499,23 @@ void Document::setHasViewTransitionPseudoElementTree(bool value)
+
+void Document::setRenderingSuppressedForViewTransition(bool value)
+{
+    if (m_renderingSuppressedForViewTransition && !value) {
```
I like writing things like this in terms of `std::exchange()`

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2384,6 +2384,15 @@ void RenderObject::RepaintRects::transform(const TransformationMatrix& matrix, f
+    // the time, when view-transitions are rare.
+    if (document().renderingSuppressedForViewTransition() && !isDocumentElementRenderer())
+        return PointerEvents::None;
+    return style().usedPointerEvents();
```
There are many style().usedPointerEvents(); calls in WebCore (some of them in some svg code). How do you tell if a newly added call should consult the renderer or the renderer's style? maybe it's always the former? -but then this PR only changes 2 of those callsites. is that sufficient?

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -828,6 +829,41 @@ void GraphicsLayerCA::setOpacity(float opacity)
+void GraphicsLayerCA::setRenderingSuppressed(bool suppressed)
+{
+    GraphicsLayer::setRenderingSuppressed(suppressed);
+    recursiveSetAncestorHasRenderingSuppression(suppressed);
```
It's weird for state to propagate to ancestors. Can't we just pick one of the root layers to set this on, and short-circuit the commits there?

mattwoodrow replied:
It's propagating to descendants. It's just so that we can short circuit painting for composited layers somewhere inside the suppressed subtree.

As you suggest below, we could short circuit `setNeedsDisplay` instead of painting itself, but we'd still need a way to know if the current layer is within a suppressed subtree (and I am using the assumption that an ancestor walk every time wouldn't be workable).

smfr commented on Source/WebCore/platform/graphics/ca/TileGrid.cpp:
```diff
@@ -713,6 +713,10 @@ void TileGrid::drawTileMapContents(CGContextRef context, CGRect layerBounds) con
 void TileGrid::platformCALayerPaintContents(PlatformCALayer* platformCALayer, GraphicsContext& context, const FloatRect&, OptionSet<GraphicsLayerPaintBehavior> layerPaintBehavior)
 {
+#if USE(CA)
+    if (containerLayer().owner() && containerLayer().owner()->platformCALayerRenderingSuppressed())
```
It feels wrong to implement rendering suppression by short circuiting painting; this makes me nervous. We should have just not make the layers as needing display.

What's the expected behavior when you resize the window while rendering is suppressed? Do we need to be able to paint the old state on layer resize?

mattwoodrow replied:
Window resizing aborts the view transition and should remove suppression so that we can repaint.

There's no expectation to be able to paint the old state again, since that would be wildly difficult.

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -1682,6 +1682,9 @@ class Document
 
     void performPendingViewTransitions();
 
+    bool renderingSuppressedForViewTransition() const;
```
This should be `isRenderingSuppressedForViewTransition()` or `renderingIsSuppressedForViewTransition()`

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -1684,6 +1684,9 @@ class Document
     void performPendingViewTransitions();
 
+    bool renderingIsSuppressedForViewTransition() const;
+    void setRenderingSuppressedForViewTransition(bool);
```
`setRenderingIsSuppressedForViewTransition`

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -2570,6 +2573,7 @@ class Document
 #endif
 
     bool m_hasViewTransitionPseudoElementTree { false };
+    bool m_renderingSuppressedForViewTransition { false };
```
`m_renderingIsSuppressedForViewTransition`

smfr replied:
The consequence of this bit getting left set by accident are catastrophic (tabs that stop rendering). I think we need some release logging or protection against botched view transitions leaving things in a bad state.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1863,7 +1863,7 @@ void Page::updateRendering()
     auto runProcessingStep = [&](RenderingUpdateStep step, const Function<void(Document&)>& perDocumentFunction) {
         m_renderingUpdateRemainingSteps.last().remove(step);
-        forEachDocument(perDocumentFunction);
+        forEachRenderableDocument(perDocumentFunction);
```
I hope using `forEachRenderableDocument` in all these places isn't a perf regression.

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -679,6 +679,9 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     bool shouldPaintUsingCompositeCopy() const { return m_shouldPaintUsingCompositeCopy; }
     void setShouldPaintUsingCompositeCopy(bool copy) { m_shouldPaintUsingCompositeCopy = copy; }
 
+    bool renderingSuppressed() const { return m_renderingSuppressed; }
```
`renderingIsSuppressed` or `isRenderingSuppressed`

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -679,6 +679,9 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     void setShouldPaintUsingCompositeCopy(bool copy) { m_shouldPaintUsingCompositeCopy = copy; }
 
+    bool renderingSuppressed() const { return m_renderingSuppressed; }
+    virtual void setRenderingSuppressed(bool suppressed) { m_renderingSuppressed = suppressed; }
```
`setRenderingIsSuppressed`

smfr replied:
This naming needs to make it clear that the state affects not just this layer, but all descendants. Maybe `setRenderingIsSuppressedIncludingDescendants()` or something.

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -828,6 +828,13 @@ void GraphicsLayerCA::setOpacity(float opacity)
+{
+    GraphicsLayer::setRenderingSuppressed(suppressed);
+    if (!suppressed)
+        setNeedsDisplay();
```
Suppression affects all descendant layers, yet this only repaints this layer which doesn't make sense.

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10530,6 +10530,22 @@ void Document::setHasViewTransitionPseudoElementTree(bool value)
+{
+    if (std::exchange(m_renderingSuppressedForViewTransition, value) && !value) {
+        if (CheckedPtr view = renderView())
+            view->compositor().setRenderingSuppressed(false);
```
I don't like that the `setRenderingSuppressed(false)` is here, but the `setRenderingSuppressed(true)` is in a function on LocalFrameView. Please factor the code to put them next to each other.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -986,6 +986,33 @@ bool LocalFrameView::flushCompositingStateForThisFrame(const LocalFrame& rootFra
+    if (!renderView->document().renderingIsSuppressedForViewTransition())
+        return;
+
+    renderView->compositor().setRenderingSuppressed(true);
```
This should be on `Document`

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -191,6 +191,8 @@ class LocalFrameView final : public FrameView {
     // Returns true if the flush was completed.
     WEBCORE_EXPORT bool flushCompositingStateIncludingSubframes();
 
+    WEBCORE_EXPORT void enableRenderingSuppressionStateIncludingSubframes();
```
I was confused about the "state" in the name, but now I realize this is "update rendering suppression based on the view transition state". Maybe it needs "View transition" in the name. Perhaps `updateRenderingSuppressionForViewTransitions()` or something.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
```diff
@@ -189,11 +189,28 @@
-    
+
+    if (owner() && owner()->platformCALayerRenderingSuppressed()) {
+        recursiveMarkWillBeDisplayed();
```
I'm confused by this. If rendering is suppressed on the PlatformCALayer (and that's just true for the root layer, right?) why are we marking the layer as "will be displayed"?

mattwoodrow committed:
[view-transitions] Suppress rendering while a view-transition is being setup.
https://bugs.webkit.org/show_bug.cgi?id=270672
rdar://124240939

Reviewed by Simon Fraser.

This implements rendering suppression by adding a flag to GraphicsLayer that
prevents updates to the PlatformCALayer tree for it and its descendants until
the flag is cleared. This will currently only work on cocoa, with remote
layers enabled.

The flag gets set on the GraphicsLayer after the rendering update completes,
so that any mutations made before rendering suppression were enabled (including
layer tree changes needed for suppression) get flushed before blocking all
future changes.

It also moves all consumers of RenderStyle:::usedPointerEvents to go via the
renderer, where we can override them to none while suppression is in place.
The exceptions are the style change handlers, which can remain unchanged
since we invalidate style entirely when the VT starts.

This doesn't yet prevent async scrolling rendering updates - bug 273612.
It also doesn't work fully with in-process layers - bug 273613.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/input-targets-root-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-css-animation-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-painting-while-render-blocked.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-raf-while-render-blocked-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setActiveViewTransition):
(WebCore::Document::renderingIsSuppressedForViewTransition const):
(WebCore::Document::setRenderingIsSuppressedForViewTransitionAfterUpdateRendering):
(WebCore::Document::clearRenderingIsSuppressedForViewTransition):
(WebCore::Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges):
(WebCore::Document::performPendingViewTransitions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::setupViewTransition):
(WebCore::ViewTransition::activateViewTransition):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::extractTargets):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::Page::doAfterUpdateRendering):
(WebCore::Page::didCompleteRenderingFrame):
(WebCore::Page::prioritizeVisibleResources):
(WebCore::Page::forEachRenderableDocument const):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::renderingIsSuppressedIncludingDescendants const):
(WebCore::GraphicsLayer::setRenderingIsSuppressedIncludingDescendants):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::needsCommit):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerRenderingIsSuppressedIncludingDescendants const):
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::unite):
(WebCore::EventRegion::unite):
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::visibleToHitTesting const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::setRenderingIsSuppressed):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::usedPointerEvents const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::checkIntersection):
(WebCore::RenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::nodeAtPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::checkIntersection):
(WebCore::LegacyRenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::nodeAtFloatPoint):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::recursiveMarkWillBeDisplayed):
(WebKit::PlatformCALayerRemote::recursiveBuildTransaction):

Canonical link: https://commits.webkit.org/279857@main
